cylc/flow/scheduler.py

```diff
@@ -1329,8 +1330,23 @@ def force_trigger_tasks(
                 f"trigger:\n  * {msg}"
             )
 
-        # remove all inactive and selected active group members
-        self.remove_matched_tasks(active_tasks_to_remove, inactive, flow_nums)
+        # Remove all inactive and selected active group members - separately
+        # because default flow inactive semantics is different for the remove
+        # command (ALL flows) then for the trigger command (ALL ACTIVE flows).
+        if flow != ["none"]:
+            # (No need to remove for no-flow triggering)
+            if active_tasks_to_remove:
+                empty_i: Set[Tuple[TaskDef, PointBase]] = set()
+                self.remove_matched_tasks(
+                    active_tasks_to_remove, empty_i, flow_nums)
+            if inactive:
+                empty_a: List[TaskProxy] = []
+                if not flow_nums:
+                    # remove for all active flows, not all flows.
+                    self.remove_matched_tasks(
+                        empty_a, inactive, self.pool.get_flow_nums([FLOW_ALL]))
+                else:
+                    self.remove_matched_tasks(empty_a, inactive, flow_nums)
```

Comment:
I thought this was right before, I'm not sure I understand this change?

Before, if `--flow=all`, then `flow_nums` is already set to all active flows:

```python
flow_nums = self.pool.get_flow_nums(flow, flow_descr)
```

But now, `flow_nums` is still all active flows, but we remove active tasks (from `flow_nums`), then we recompute the `flow_nums`:

```python
self.pool.get_flow_nums([FLOW_ALL])
```

Which may now differ from `flow_nums` because we haven't spawned the new task(s) yet?

Reply:
> I thought this was right before, I'm not sure I understand this change? 

I'll double check this soon, but I think what you've missed is flow_nums can be *empty*.  For trigger that means "all active" flows, but for remove that means "all" flows. That's why I'm recomputing them with an explict FLOW_ALL on calling the remove method.

Reply:
(I've spent all on day #6813  and #6821 - back to this tomorrow).

Reply:
OK I finally got back to this. 

Sorry, my bad - I think I missed your main point because I didn't notice that I'd stupidly moved the `get_flow_nums` call into the remove call arg list to avoid a temporary variable ... by jumping over the earlier remove call that could change the task pool.

I'll try to clean up flow-nums handling a bit. It's gotten quite confusing as now have three commands (set, trigger, remove) that each handle the same CLI options slightly differently.

Reply:
I'm resolving this discussion, after significantly refactoring flow-number handling to allow (among other things) simplifying the remove-via-trigger bit. (We can discuss the new code if you want to query anything there).